src/llama.cpp

```diff
@@ -2092,6 +2112,11 @@ struct llama_hparams {
     uint32_t n_expert_used = 0;
     uint32_t n_vocab_type = 0; // for BERT-style token types
 
+    // TODO: find a more compact way to add more per-layer hyper-parameters
+    std::vector<int32_t> n_head_vec;
+    std::vector<int32_t> n_head_kv_vec;
+    std::vector<int32_t> n_ff_vec;
```

Comment:
Do we expect negative values here? If not, we should change to `uint32_t`

Reply:
I've used `int32_t` here due to a limitation of `GGUFWriter`, which doesn't allow choosing the specific type of integer arrays and defaults to `INT32`. And also because `get_arr` in `src/llama.cpp` can only load arrays into vectors which have exactly the same type as the loaded array.

I'll see if this can be fixed, but it will likely require using Numpy types (e.g. `np.uint32`) in `GGUFValueType.get_type()` in `gguf-py/gguf/constants.py`.